Bottom: b80606b56c3c2b5130bf3325a8e435daa58f0f9a
Top:    17760921d8eef8dbcf3148e8ac062e2b60ca482d
Author: Peter Clifton <pcjc2@cam.ac.uk>
Date:   2009-11-12 01:24:49 +0000

Correct rendering of and connectivity checks for zero clearance pads and pins

NB: These areren't technically allowed by PCB, but it is nice that when a
user hacks zero clearance in their PCB file, that we:

a) Draw polygons without any cleared gaps in the output
b) Correctly determine that these objects will be connected to the polygon


---

diff --git a/src/draw.c b/src/draw.c
index 4657f7c..4d7c8de 100644
--- a/src/draw.c
+++ b/src/draw.c
@@ -1365,6 +1365,9 @@ DrawPadLowLevel (hidGC gc, PadTypePtr Pad, Boolean clear, Boolean mask)
       return;
     }
 
+  if (clear && !mask && Pad->Clearance <= 0)
+    return;
+
   if (TEST_FLAG (THINDRAWFLAG, PCB) ||
       (clear && TEST_FLAG (THINDRAWPOLYFLAG, PCB)))
     {
diff --git a/src/find.c b/src/find.c
index 6e81d29..30a1ea5 100644
--- a/src/find.c
+++ b/src/find.c
@@ -2288,7 +2288,8 @@ LOCtoPadPoly_callback (const BoxType * b, void *cl)
   struct lo_info *i = (struct lo_info *) cl;
 
 
-  if (!TEST_FLAG (TheFlag, polygon) && !TEST_FLAG (CLEARPOLYFLAG, polygon))
+  if (!TEST_FLAG (TheFlag, polygon) &&
+      (!TEST_FLAG (CLEARPOLYFLAG, polygon) || !i->pad.Clearance))
     {
       if (IsPadInPolygon (&i->pad, polygon) &&
           ADD_POLYGON_TO_LIST (i->layer, polygon))
@@ -2529,15 +2530,12 @@ LookupLOConnectionsToPolygon (PolygonTypePtr Polygon, Cardinal LayerGroup)
         }
       else
         {
-          if (!TEST_FLAG (CLEARPOLYFLAG, Polygon))
-            {
-              info.layer = layer - max_layer;
-              if (setjmp (info.env) == 0)
-                r_search (PCB->Data->pad_tree, (BoxType *) & info.polygon,
-                          NULL, LOCtoPolyPad_callback, &info);
-              else
-                return True;
-            }
+          info.layer = layer - max_layer;
+          if (setjmp (info.env) == 0)
+            r_search (PCB->Data->pad_tree, (BoxType *) & info.polygon,
+                      NULL, LOCtoPolyPad_callback, &info);
+          else
+            return True;
         }
     }
   return (False);
diff --git a/src/polygon.c b/src/polygon.c
index ec53c70..8bf8049 100644
--- a/src/polygon.c
+++ b/src/polygon.c
@@ -754,6 +754,8 @@ SubtractPad (PadType * pad, PolygonType * p)
 {
   POLYAREA *np = NULL;
 
+  if (pad->Clearance == 0)
+    return 0;
   if (TEST_FLAG (SQUAREFLAG, pad))
     {
       if (!
@@ -861,6 +863,8 @@ pad_sub_callback (const BoxType * b, void *cl)
   /* don't subtract the object that was put back! */
   if (b == info->other)
     return 0;
+  if (pad->Clearance == 0)
+    return 0;
   polygon = info->polygon;
   if (XOR (TEST_FLAG (ONSOLDERFLAG, pad), !info->solder))
     {
